### PR TITLE
DEPR: DataFrameGroupBy.apply operating on the group keys

### LINKED ISSUE #7155: API: way to exclude the grouped column with apply
http://stackoverflow.com/questions/23709811/best-way-to-sum-group-value-counts-in-pandas/23712433?noredirect=1#comment36441762_23712433
any of these look palatable?

`df.groupby('c',as_index='exclude').apply(....)`
`df.groupby('c')['~c'].apply(...)`
`df.groupby('c',filter='c')`
`df.groupby('c',filter=True)`

rather than doing a negated selection (`df.columns-['c']`)

```
df = pd.DataFrame.from_dict({'b1':['aa','ac','ac','ad'],\
                             'b2':['bb','bc','ad','cd'],\
                             'b3':['cc','cd','cc','ae'],\
                             'c' :['a','b','a','b']})

df.groupby('c')[df.columns - ['c']].apply(lambda x: x.unstack().value_counts())
Out[92]: 
c    
a  cc    2
   bb    1
   ad    1
   ac    1
   aa    1
b  cd    2
   ad    1
   ae    1
   ac    1
   bc    1
dtype: int64
```

### PR DESCRIPTION
**Pull request by rhshadrach**
- [x] closes #7155 (Replace xxxx with the GitHub issue number)
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

**Comment by topper-123** [comment]
I will look in depth in a few hours, I got some obligations. Just initially: 

Is it correct that this will help simplify the code paths for other groupby methods or is this a change that only will affect groupby `apply` itself? If yes, could you explain how the other groupby methods use/are used by `apply`?

**Comment by rhshadrach** [comment]
> Is it correct that this will help simplify the code paths for other groupby methods or is this a change that only will affect groupby `apply` itself? 

This change is only for apply itself. Ignoring filtrations (which by design include the grouping columns when applicable), almost all of groupby does not operate on the grouping columns. apply is the major exception, and this deprecation is to correct that.

> If yes, could you explain how the other groupby methods use/are used by `apply`?

Almost no groupby methods use `apply` internally. Those that do:

 - first, last
 - value_counts
 - is_monotonic_increasing, is_monotonic_decreasing
 - dtype
 - diff (for axis=1 only)

**Comment by topper-123** [comment]
So this can't affect `SeriesGroupby` at all and no `DataFrameGroupby` method uses `apply`. So that means basically no worries wrt. effect on other method, which is great.

I've looked at the code in `Groupby.apply` and you did in the `if` clause: `self._selected_obj.shape != self._obj_with_exclusions.shape`, which surprised me. So I tried:

```python
>>> df = pd.DataFrame({"A": ["a", "a", None, None, "b", "b", "c"], "B": [1, 1, 2, 2, 3, 3, 1]})
>>> g = df.groupby('A')[df.columns]
>>> g.sum()  # sums over grouping column
   A    B
A
1  3  aac
2  4    0
3  6   bb
```

I never knew this was possible (works also with `apply`), but I think this is a good thing: by default let's exclude grouping columns from `apply`/`agg` etc., but this let's users get them back in in the rare case they need that (there's always an edge  case...).

Do you know if this is documented anywhere? IMO we should officially support that pattern with docs and tests. Probably also show this as an example in the deprecation text?

File: pandas/tests/resample/test_datetime_index.py [hunk 1]
```diff
@@ -1039,8 +1039,12 @@ def test_resample_segfault(unit):
         all_wins_and_wagers, columns=("ID", "timestamp", "A", "B")
     ).set_index("timestamp")
     df.index = df.index.as_unit(unit)
-    result = df.groupby("ID").resample("5min").sum()
-    expected = df.groupby("ID").apply(lambda x: x.resample("5min").sum())
+    msg = "DataFrameGroupBy.resample operated on the grouping columns"
+    with tm.assert_produces_warning(FutureWarning, match=msg):
+        result = df.groupby("ID").resample("5min").sum()
```
Comment (topper-123) [inline pandas/tests/resample/test_datetime_index.py:1044]: `resample` gives a warning? can we avoid that? (Does it operate on the grouping columns?)

**Review by topper-123 — COMMENTED** [review]
A few comments/questions.

Can you especially look into if we can avoid giving the deprecation warning in `groupby.resample`?

File: pandas/core/groupby/groupby.py [hunk 2]
```diff
@@ -1487,6 +1487,23 @@ def f(g):
...
+                    not isinstance(self.obj, Series)
+                    and self._selection is None
+                    and self._selected_obj.shape != self._obj_with_exclusions.shape
+                ):
+                    msg = (
+                        f"{type(self).__name__}.apply operated on the grouping "
+                        f"columns. This behavior is deprecated, and in a future "
+                        f"version of pandas the grouping columns will be excluded "
+                        f"from the operation. Subset the data to exclude the "
+                        f"groupings and silence this warning."
```
[the remark this answers is not in the document]
Comment (rhshadrach) [reply pandas/core/groupby/groupby.py:1500]: What's the benefit of removing the phrase "to exclude the groupings"?

[the remark this answers is not in the document]
Comment (topper-123) [reply pandas/core/groupby/groupby.py:1500]: My idea was that by keeping  the grouping columns in the subsetting, the users are guaranteed to get the same rseult as before, but without the warning:

```python
>>> df = pd.DataFrame({"A": [1, 1, 2, 2], "B": [1,  2, 3, 4]})
>>> g = df.groupby('A')
>>> g.apply(lambda x: x.sum())
   A  B
A
1  2  3
2  4  7
>>> g[df.columns].apply(lambda x: x.sum())
   A  B
A
1  2  3
2  4  7
```

Or they may actually not want to include, then remove it from the subsetting,

```python
>>> g[df.columns.drop("A")].apply(lambda x: x.sum())
   B
A
1  3
2  7
```

the idea is just that they may/may not want to remove the groupings in the subset and "exclude the  groupings" may not be what they want in all cases.

File: pandas/tests/resample/test_datetime_index.py (hunk 1 above)
Comment (topper-123) [reply pandas/tests/resample/test_datetime_index.py:1044]: Then it should be deprecated and mentioned in the whatsnew also, right?

File: pandas/core/groupby/groupby.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (rhshadrach) [reply pandas/core/groupby/groupby.py]: I see - this makes sense, but I find it confusing to call this "subsetting". I'll see what I can do for the warning message here.

File: doc/source/whatsnew/v2.1.0.rst [hunk 3]
```diff
@@ -149,6 +149,7 @@ Other API changes
 
 Deprecations
 ~~~~~~~~~~~~
+- Deprecated :meth:`.DataFrameGroupBy.apply` operating on the grouping column(s) (:issue:`7155`)
```
[the remark this answers is not in the document]
Comment (rhshadrach) [reply doc/source/whatsnew/v2.1.0.rst]: Thanks - done.

File: pandas/core/resample.py [hunk 4]
```diff
@@ -1344,7 +1365,18 @@ def func(x):
 
             return x.apply(f, *args, **kwargs)
 
-        result = self._groupby.apply(func)
+        msg = (
+            "DataFrameGroupBy.resample operated on the grouping columns. "
+            "This behavior is deprecated, and in a future version of "
+            "pandas the grouping columns will be excluded from the operation. "
+            "Subset the data to exclude the groupings and silence this warning."
```
[the remark this answers is not in the document]
Comment (rhshadrach) [reply pandas/core/resample.py:1368]: Done.

File: pandas/tests/resample/test_datetime_index.py (hunk 1 above)
Comment (rhshadrach) [reply pandas/tests/resample/test_datetime_index.py:1044]: Ah, yes, I missed mentioning it there. Added.

File: pandas/core/groupby/groupby.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (rhshadrach) [reply pandas/core/groupby/groupby.py]: Updated. What do you think?

**Comment by rhshadrach** [comment]
> Do you know if this is documented anywhere? IMO we should officially support that pattern with docs and tests. 

Added a line in the user guide; I know we test this because that's how I discovered it but I'll have to dig up where that is.

> Probably also show this as an example in the deprecation text for this change?

In the whatsnew, we could add a notable deprecations section to do this, but I'm not sure it's worth it. I'm rather against having an example in the FutureWarning that is raised.

File: doc/source/user_guide/groupby.rst [hunk 5]
```diff
@@ -1067,7 +1067,7 @@ missing values with the ``ffill()`` method.
    ).set_index("date")
    df_re
 
-   df_re.groupby("group").resample("1D").ffill()
+   df_re.groupby("group")[["val"]].resample("1D").ffill()
```
Comment (topper-123) [inline doc/source/user_guide/groupby.rst:1076]: This is different than the original code.

In this example the user probably want to keep "group" column, so I'd do`[["group", "val"]]`.

File: doc/source/whatsnew/v2.1.0.rst [hunk 6]
```diff
@@ -149,7 +149,7 @@ Other API changes
 
 Deprecations
 ~~~~~~~~~~~~
-- Deprecated :meth:`.DataFrameGroupBy.apply` operating on the grouping column(s) (:issue:`7155`)
+- Deprecated :meth:`.DataFrameGroupBy.apply` and :class:`.Resampler` methods operating on the grouping column(s); subset the data to exclude the grouping column(s) to adopt future behavior and avoid the ``FutureWarning`` (:issue:`7155`)
```
Comment (topper-123) [inline doc/source/whatsnew/v2.1.0.rst]: This PR has no affect `Resampler` itself, right?

I'm not sure I would understand the phrase "and :class:`.Resampler` methods operating on the grouping column(s)". Instead maybe "and the methods in :meth:`DataFrameGroupBy.resample`  operating on the grouping column(s).

File: doc/source/user_guide/groupby.rst (hunk 5 above)
Comment (rhshadrach) [reply doc/source/user_guide/groupby.rst:1076]: Why would you want to have the group in both the index and the columns?

File: doc/source/whatsnew/v2.1.0.rst (hunk 6 above)
Comment (rhshadrach) [reply doc/source/whatsnew/v2.1.0.rst]: > This PR has no affect `Resampler` itself, right?

True, it's technically DatetimeIndexResamplerGroupby, PeriodIndexResamplerGroupby, and TimedeltaIndexResamplerGroupby. 

> the methods in :meth:DataFrameGroupBy.resample

I think "the methods on the objects returned by :meth:DataFrameGroupBy.resample...

**Review by topper-123 — APPROVED** [review]
Nice, LGTM..

**Review by mroeschke — CHANGES_REQUESTED** [review]
Looks good. Some of the doctest look to need updating though

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks @rhshadrach

**Comment by rhshadrach** [comment]
The warning should be triggered only when both of the following apply:

 - The UDF with the grouping columns does not raise
 - Raising would mean the fallback passes different groups (namely, excluding the grouping columns)

The only thing I could see doing in addition is trying (perhaps on just the first group) if the result differs whether the grouping columns are included or not. What do you think about this option?

Another option would be changing without deprecation. I'm also planning to put up a PDEP to propose a bunch of changes to agg/apply/transform that get put up behind an option for smoother transition. Perhaps this could go behind that option as well.

**Comment by phofl** [comment]
If I understand you here, the default case would raise a warning:

```
df = pd.DataFrame({"a": [1, 2, 3], "b": 1.5, "c": True, "d": "d"})


def test(x):
    return x


df.groupby("d").apply(test)
```

which it does right now. This seems pretty noisy, since there is no way to get rid of the warning as a user apart from changing your code everywhere?

I think an option is a good idea, if you want to add something like this anyway for other reasons.

**Comment by rhshadrach** [comment]
> This seems pretty noisy

I would typically describe "noisy" as warning when behavior won't change, so just want to be clear that this will.

```
# Current
     a    b     c  d
d                   
d 0  1  1.5  True  d
  1  2  1.5  True  d
  2  3  1.5  True  d

# Future
     a    b     c
d                   
d 0  1  1.5  True
  1  2  1.5  True
  2  3  1.5  True
```

To suppress the warning and adopt future behavior, one would do `df.groupby("d")[["a", "b", "c"]].apply(test)`; to keep current behavior it's `df.groupby("d")[["a", "b", "c", "d"]].apply(test)`. Neither of these warn.

But agreed it's noisy in the sense of "someone using apply a lot of may see a lot of warnings" - and experience a lot of behavior changes!

I'll put up a PR to revert.

**Comment by rhshadrach** [comment]
> Values are constant anyway, so why would you need the information? That's what I mean with noisy. I am assuming that a vast majority of users just doesn't care about the grouping columns

It sounds like you're saying you'd be okay changing this behavior without deprecation?

The users may not want them, but they have them now. So if their workflow drops the grouping columns from the result, that workflow will break if we were to change this behavior without deprecation.

**Comment by phofl** [comment]
> I'm also planning to put up a PDEP to propose a bunch of changes to agg/apply/transform that get put up behind an option for smoother transition

I'd prefer this solution if you end up creating an option anyway. 

But I think I'd also be ok with doing this without a deprecation. Personally, I'd be annoyed that I'd have to change all my apply calls just to silence the warning even though I don't care about the grouping columns

**Comment by rhshadrach** [comment]
> > I'm also planning to put up a PDEP to propose a bunch of changes to agg/apply/transform that get put up behind an option for smoother transition
> 
> I'd prefer this solution if you end up creating an option anyway.

We've moved away from introducing this option. I'm now looking for ways to introduce this deprecation again.

> But I think I'd also be ok with doing this without a deprecation. Personally, I'd be annoyed that I'd have to change all my apply calls just to silence the warning even though I don't care about the grouping columns

I imagine this is a popular method and we're changing the output in what I fear could be a common case. Doing this without deprecation seems too likely to break a lot of code. I don't fully understand your comment about "changing apply calls just to silence the warning". Wouldn't your code have a good chance of breaking if you didn't change it ahead of time when we change the behavior?

**Comment by phofl** [comment]
> I imagine this is a popular method and we're changing the output in what I fear could be a common case. Doing this without deprecation seems too likely to break a lot of code. I don't fully understand your comment about "changing apply calls just to silence the warning". Wouldn't your code have a good chance of breaking if you didn't change it ahead of time when we change the behavior?

Not really if you don't care about the group keys being included in your group or not. This was often the case for me when I used apply.

But I agree, not deprecating this in some way is probably a bad idea.

**Comment by rhshadrach** [comment]
I believe the checks on the warning that were included here completely predict when the output would change. In other words, we have no way of making this less noisy by merely emitting warnings in less cases.

What do you think about an option to adopt the future behavior? Something like `mode.groupings_in_apply_result`. Defaults to `True` (current behavior). When set to `False`, it would adopt the future behavior. Then we deprecate the option in 3.x. Is that palatable?

If that's a good way forward, I think we still make the changes to the tests as they are in this PR.

**Comment by topper-123** [comment]
Could an idea be to silence the deprecation warning until after 2.1 or maybe even 2.2?

IMO we need a deprecation here, but it may not be necessary to have it effective in 2.1, especially if silencing/unsilencing is low effort?

**Comment by rhshadrach** [comment]
I'm fine with waiting to deprecate; but I think to @phofl's point, silencing is not low effort (especially when you have many columns you want to include)

**Comment by topper-123** [comment]
Sorry, I meant if we in the `pandas/core/groupby/groupby.py` surround the FutureWarning with `warnings.catch_warnings():` and `warnings.filter_warning` code, then remove that code when we want to re-activate the warnings e.g. in v2.2.

**Comment by rhshadrach** [comment]
@topper-123: I'd personally prefer just to wait until 2.1 is released. When you wrote:

> especially if silencing/unsilencing is low effort?

I thought you meant think from a _user_ perspective. From an implementation perspective, almost all the work is with the tests. But I'm okay with having to redo a little work if the tests have changes from the implementation here by the time 2.1 is release.

**Comment by topper-123** [comment]
Ok, sounds good for me.

**Pull request merged**